### PR TITLE
support for shadow ioeventfd

### PR DESCRIPTION
**Pull request by tmakatos**
Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

**Comment by jlevon** [comment]
should be make this a config option, as it's not something officially supported by anything else?

File: include/libvfio-user.h [hunk 1]
```diff
@@ -1073,7 +1073,7 @@ vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg);
 int
 vfu_create_ioeventfd(vfu_ctx_t *vfu_ctx, uint32_t region_idx, int fd,
                      size_t offset, uint32_t size, uint32_t flags,
-                     uint64_t datamatch);
+                     uint64_t datamatch, int data_fd);
```
[the remark this answers is not in the document]
Comment (jlevon) [reply include/libvfio-user.h:1076]: is "data_fd" the best name too?

File: lib/libvfio-user.c [hunk 2]
```diff
@@ -1832,6 +1840,10 @@ vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
 
     assert(vfu_ctx != NULL);
 
+    /*
```
Comment (jlevon) [inline lib/libvfio-user.c:1843]: make this an spec issue instead?

**Review by jlevon — COMMENTED** [review]
should add some py tests too?

**Comment by tmakatos** [comment]
My primary goal was to get some initial feeback, I think until we hear back from ElenaU about ioregionfd we should leave this unmerged.

File: include/libvfio-user.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmakatos) [reply include/libvfio-user.h:1076]: I'm open to suggestions, how about `shadow_fd`?

File: lib/libvfio-user.c (hunk 2 above)
Comment (tmakatos) [reply lib/libvfio-user.c:1843]: sure

**Comment by tmakatos** [comment]
> should be make this a config option, as it's not something officially supported by anything else?

sure

File: include/libvfio-user.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlevon) [reply include/libvfio-user.h:1076]: I like it!

File: lib/libvfio-user.c (hunk 2 above)
Comment (tmakatos) [reply lib/libvfio-user.c:1843]: https://github.com/nutanix/libvfio-user/issues/699

**Comment by tmakatos** [comment]
>

**Comment by tmakatos** [comment]
> should be make this a config option, as it's not something officially supported by anything else?

done

**Comment by tmakatos** [comment]
> should add some py tests too?

yep I forgot

**Comment by tmakatos** [comment]
> > should add some py tests too?
> 
> yep I forgot

I'm not sure how to add _new_ tests because we rely on a `#define` at build time to enable support for shadow ioeventfd, do you know if this is straightforward to do in meson? Alternatively, we always include support for shadow ioeventfd and enable it during runtime differently, e.g. via an environment variable or setting some flag when creating the context? Or would that be too dangerous?

**Comment by jlevon** [comment]
> > > should add some py tests too?
> > 
> > 
> > yep I forgot
> 
> I'm not sure how to add _new_ tests because we rely on a `#define` at build time to enable support for shadow ioeventfd, do you know if this is straightforward to do in meson? Alternatively, we could always include support for shadow ioeventfd and enable it during runtime differently, e.g. via an environment variable or setting some flag when creating the context? Or would that be too dangerous?

It should be a meson option, that then sets a define. then you can look at the meson option to decide whether to run those test cases.

https://mesonbuild.com/Build-options.html

**Comment by jlevon** [comment]
will review when tests are passing

File: include/libvfio-user.h [hunk 3]
```diff
@@ -1069,12 +1069,18 @@ vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg);
  * @size: size of the ioeventfd
  * @flags: Any flags to set up the ioeventfd
  * @datamatch: sets the datamatch value
+ * @shadow_fd: File descriptor than can be mmap'ed, KVM will write there the
```
Comment (jlevon) [inline include/libvfio-user.h:1072]: that

File: include/libvfio-user.h [hunk 4]
```diff
@@ -1069,12 +1069,18 @@ vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg);
...
+ *  otherwise discarded value when the ioeventfd is written to. If set to -1
+ *  then a normal ioeventfd is set up instead of a shadow one. Then vfio-user
```
Comment (jlevon) [inline include/libvfio-user.h:1074]: s/Then/The/

File: lib/libvfio-user.c [hunk 5]
```diff
@@ -555,6 +562,7 @@ handle_device_get_region_io_fds(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
     ioeventfd_t *sub_reg = NULL;
     size_t nr_sub_reg = 0;
     size_t i = 0;
+    size_t nr_shdw_reg = 0;
```
Comment (jlevon) [inline lib/libvfio-user.c:565]: nit, s/shdw/shadow/ ? The shorter version doesn't seem better here

File: test/py/meson.build [hunk 6]
```diff
@@ -49,6 +49,11 @@ python_tests = [
     'test_vfu_realize_ctx.py',
 ]
 
+opt_shadow_ioeventfd = get_option('shadow-ioeventfd')
```
Comment (jlevon) [inline test/py/meson.build:52]: any need for this temporary variable?

File: test/py/test_shadow_ioeventfd.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    # vfio-user app reads the eventfd, there should be nothing there
+    try:
+        os.read(efd, IOEVENT_SIZE)
+    except BlockingIOError as e:
+        if e.errno != errno.EAGAIN:
+            assert False
+    else:
+        assert False
+
+    # Client writes to the I/O region. Tthe write to the eventfd would by done
```
Comment (jlevon) [inline test/py/test_shadow_ioeventfd.py:94]: s/Tthe/The
s/by/be

**Review by jlevon — COMMENTED** [review]
Looks good, just nits. Mind adding a short explanation to docs/ too ?

**Comment by tmakatos** [comment]
> Looks good, just nits. Mind adding a short explanation to docs/ too ?

sure

File: include/libvfio-user.h (hunk 3 above)
Comment (tmakatos) [reply include/libvfio-user.h:1072]: done

File: lib/libvfio-user.c (hunk 5 above)
Comment (tmakatos) [reply lib/libvfio-user.c:565]: ok

File: include/libvfio-user.h (hunk 4 above)
Comment (tmakatos) [reply include/libvfio-user.h:1074]: done

File: test/py/meson.build (hunk 6 above)
Comment (tmakatos) [reply test/py/meson.build:52]: nope

File: docs/ioregionfd.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+from ioctl(KVM_RUN).".
+
+libvfio-user currently supports an experimental variant of this mechanism
+called shadow ioeventfd. A shadow ioeventfd is a normal ioeventfd where the
+vfio-user server passes another piece of memory (called the _shadow_ memory)
+via an additional file descriptor when configuring the ioregionfd, which then
+QEMU memory maps and passes this address to KVM. This shadow memory is never
+exposed to the guest. When the guest writes to the trapped memory, KVM writes
+the value to the shadow memory instread of discarding it, and then proceeds
+kicking the evetnfd as normal.
```
Comment (jlevon) [inline docs/ioregionfd.md:16]: "eventfd"

Can we put links to the relevant qemu+SPDK branches, with some brief instructions on how to use? Perhaps later?

Comment (tmakatos) [reply docs/ioregionfd.md:16]: sure

**Review by jlevon — APPROVED** [review]
thanks

**Pull request merged**